File: bench/nqueens/nqueens_pure.py
"""N-Queens: pure immutable approach — list[:] slice copy per placement (same as Sprout)."""
import sys, time
sys.setrecursionlimit(5000)

def queens(n, row, col, cols, pd, nd):
    if row == n: return 1
    if col >= n: return 0
    skip = queens(n, row, col + 1, cols, pd, nd)
    p = row + col
    q = row - col + n - 1
    if not cols[col] and not pd[p] and not nd[q]:
        c2 = cols[:]; c2[col] = True
        p2 = pd[:];   p2[p]   = True
        n2 = nd[:];   n2[q]   = True
        place = queens(n, row + 1, 0, c2, p2, n2)
    else:
        place = 0
    return skip + place

def count_solutions(n):
    return queens(n, 0, 0, [False]*n, [False]*(2*n-1), [False]*(2*n-1))

for n in [1, 4, 8, 10, 12, 13]:
    t = time.perf_counter()
    c = count_solutions(n)
    ms = (time.perf_counter() - t) * 1000
    print(f"N={n}: {c}  ({ms:.1f} ms)")

File: bench/nqueens/nqueens_pure.rb
# N-Queens: pure immutable approach — Array#dup per placement (same as Sprout).
# Run: ruby nqueens_pure.rb

def queens(n, row, col, cols, pos_diag, neg_diag)
  return 1 if row == n
  return 0 if col >= n

  skip = queens(n, row, col + 1, cols, pos_diag, neg_diag)

  pd = row + col
  nd = row - col + n - 1

  if !cols[col] && !pos_diag[pd] && !neg_diag[nd]
    c2 = cols.dup;     c2[col] = true
    p2 = pos_diag.dup; p2[pd]  = true
    n2 = neg_diag.dup; n2[nd]  = true
    place = queens(n, row + 1, 0, c2, p2, n2)
  else
    place = 0
  end

  skip + place
end

def count_solutions(n)
  queens(n, 0, 0,
    Array.new(n,       false),
    Array.new(2*n - 1, false),
    Array.new(2*n - 1, false))
end

[1, 4, 8, 10, 12, 13].each do |n|
  t = Process.clock_gettime(Process::CLOCK_MONOTONIC)
  c = count_solutions(n)
  ms = (Process.clock_gettime(Process::CLOCK_MONOTONIC) - t) * 1000
  printf "N=%d: %d  (%.1f ms)\n", n, c, ms
end
